lib/mri/uri/generic.rb

```diff
@@ -591,7 +591,7 @@ def check_host(v)
       if @opaque
         raise InvalidURIError,
           "can not set host with registry or opaque"
-      elsif parser.regexp[:HOST] !~ v
+      elsif !parser.regexp[:HOST].match v
```

Comment:
I was hoping to properly fix `Regexp.=~` to not deoptimise when setting the results of named matches, but that's a fairly hairy bit of work to do in the translator (because it should only set variables if the regexp is a constant) so I put this in as a stopgap.